catlas/filter_utils.py

```diff
@@ -348,3 +350,301 @@ def get_first_type(x):
+    Estimates surface density multiplied by cohesive energy
+    Args:
+        ucell (Structure): PMG Structure representation of a bulk unit cell.
+        slab (Slab): PMG Structure representation of a Slab cell.
+        ecoh (float): Cohesive energy which correlates to the surface energy
+        factor (float): buffer for the radius to look
+            for neighbors in order to calculate bond length
+
+    Returns:
+        (float): Rough estimate of cohesive energy x surface density
+    """
+    slab = row["slab_structure"]
+    ucell = row["bulk_structure"]
+    a = surface_area(slab)
+    cns = get_total_nn(ucell, slab, neighbor_factor)
+    return cns * (1 / (2 * a))
+
+
+def filter_by_surface_property(bag_partition, name: str, val: dict):
+    """
+    Parse all miller facets per material and pick those that should be lower energy
+    by the broken bond model or the surface density model
+
+    Args:
+        bag_partition (Iterable[dict]: a partition of a dask bag containing enumerated surfaces
+        name: filter name to be applied (comes from the main catlas config)
+        val: values associated with name from the config yaml file, which futher
+            specify how the filter shoulf be applied
+    Returns:
+        Iterable[dict]: the bag partition with undesired slabs filtered out
+    """
+    # Use either the provided hashes, or default to the surface atoms object
+    hash_column = "bulk_id"
+
+    neighbor_factor = val["neighbor_factor"] if "neighbor_factor" in val else 1.1
+
+    # Hash all entries by the desired columns
+    hash_dict = {}
+    for row in bag_partition:
+        key = row[hash_column]
+        if key in hash_dict:
+            hash_dict[key].append(row)
+        else:
+            hash_dict[key] = [row]
+
+    # Iterate over all unique hashes
+    filtered_bag_partition = []
+    for key, value in hash_dict.items():
+        if name == "filter_by_broken_bonds":
+            surface_model_values = [
+                get_broken_bonds(row, neighbor_factor) for row in value
+            ]
+        elif name == "filter_by_surface_density":
+            surface_model_values = [
+                get_surface_density(row, neighbor_factor) for row in value
+            ]
+
+        if "top_k" in val:
+            select_indices = np.argpartition(surface_model_values, val["top_k"])[
+                : val["top_k"]
+            ]
+
+        elif "top_proportion" in val:
+            k = int(np.ceil(val["top_proportion"] * len(surface_model_values)))
+            select_indices = np.argpartition(surface_model_values, k)[:k]
+
+        filtered_bag_partition.extend(
+            [row for idx, row in enumerate(value) if idx in select_indices]
+        )
+
+    return filtered_bag_partition
+
+
+def filter_best_facet_by_surface_property(bag_partition, name: str, val: dict):
+    """
+    Parse each facet and pick the one that should be the lowest energy
+    by the broken bond model or the surface density model
```

Comment:
Why newline and no indentation?

Reply:
I think that only applies to args/returns/raises